tools/checked-c-convert/CheckedCConvert.cpp

```diff
@@ -732,64 +732,84 @@ bool CastPlacementVisitor::VisitFunctionDecl(FunctionDecl *FD) {
-					didAny = true;
-				} else {
-					// Do what we used to do. 
-					std::string v = Decl->mkString(Info.getConstraints().getVariables());
-					parmStrs.push_back(v);
-				}
-			}
-
-		// Compare returns. 
-		auto Decl = getHighest(cDecl->getReturnVars(), Info);
-		auto Defn = getHighest(cDefn->getReturnVars(), Info);
-
-		// Insert a bounds safe interface at the return address. 
-		std::string returnVar = "";
-		std::string endStuff = "";
-		if (Defn->isLt(*Decl, Info)) {
-			std::string ctype = Defn->mkString(Info.getConstraints().getVariables());
-			returnVar = Defn->getTy();
-			endStuff = " : itype("+ctype+") ";
-			didAny = true;
-		} else {
-			// Do what we used to do at the return address. 
-			returnVar = Decl->mkString(Info.getConstraints().getVariables()) + " ";
-		}
-
-		s = returnVar + cDecl->getName() + "(";
-		if (parmStrs.size() > 0) {
-			std::ostringstream ss;
-
-			std::copy(parmStrs.begin(), parmStrs.end() - 1,
-					 std::ostream_iterator<std::string>(ss, ", "));
-			ss << parmStrs.back();
-
-			s = s + ss.str() + ")";
-		} else {
-			s = s + "void)";
-		}
-
-		if (endStuff.size() > 0)
-			s = s + endStuff;
+    std::string s = "";
+    std::vector<std::string> parmStrs;
+    // Compare parameters. 
+    if (cDecl->numParams() == cDefn->numParams())
+      for (unsigned i = 0; i < cDecl->numParams(); ++i) {
+        auto Decl = getHighest(cDecl->getParamVar(i), Info);
+        auto Defn = getHighest(cDefn->getParamVar(i), Info);
+        assert(Decl);
+        assert(Defn);
+
+        didAny = true;
+        // If this holds, then we want to insert a bounds safe interface. 
+        bool anyConstrained = Defn->anyChanges(Info.getConstraints().getVariables());
+        if (Defn->isLt(*Decl, Info) && anyConstrained) {
+          std::string scratch = "";
+          raw_string_ostream declText(scratch);
+          Definition->getParamDecl(i)->print(declText);
+          std::string ctype = Defn->mkString(Info.getConstraints().getVariables(), false);
+          std::string bi = declText.str() + " : itype("+ctype+") ";
+          parmStrs.push_back(bi);
+        } else {
+          // Do what we used to do.
+          if (anyConstrained) { 
+            std::string v = Defn->mkString(Info.getConstraints().getVariables());
+            if (PVConstraint *PVC = dyn_cast<PVConstraint>(Defn)) {
+              if (PVC->getItypePresent()) {
+                v = v + " : " + PVC->getItype();
+              }
+            }
+            parmStrs.push_back(v);
+          } else {
+            std::string scratch = "";
+            raw_string_ostream declText(scratch);
+            Definition->getParamDecl(i)->print(declText);
+            parmStrs.push_back(declText.str());
+                      }
+        }
+      }
+
+    // Compare returns. 
+    auto Decl = getHighest(cDecl->getReturnVars(), Info);
+    auto Defn = getHighest(cDefn->getReturnVars(), Info);
+
+    // Insert a bounds safe interface at the return address. 
```

Comment:
Instead of saying "at the return address", should the comment say "insert a bounds-safe interface for the return"?

Reply:
Yeah, probably.